### PR TITLE
Support for automatic unlinking in failure cases

### PR DESCRIPTION
**Pull request by darrylmasson**
Supporting https://github.com/XENONnT/daqnt/pull/68

### CODE REVIEW

**Comment by darrylmasson** [comment]
Additionally, this PR fixes #66

File: dispatcher/MongoConnect.py [hunk 1]
```diff
@@ -203,6 +198,7 @@ def aggregate_status(self):
                     'pll_unlocks': 0,
                     'number': -1}
                 for k in self.dc}
+        phys_stat = {}
```
Comment (jorana) [inline dispatcher/MongoConnect.py:201]:
```suggestion
        phys_stat = defaultdict(list)  # list of statussus of each of the phys. detectors
```

File: dispatcher/MongoConnect.py [hunk 2]
```diff
@@ -220,69 +218,38 @@ def aggregate_status(self):
...
-
+                status = self.extract_status(doc, now_time)
                 statuses[doc['host']] = status
+                doc['status'] = status
                 modes.append(doc.get('mode', 'none'))
                 run_nums.append(doc.get('number', None))
                 aggstat[phys_det]['status'] = status
                 aggstat[phys_det]['mode'] = modes[-1]
                 aggstat[phys_det]['number'] = run_nums[-1]
+                phys_stat[phys_det].append(status)
```
Comment (jorana) [inline dispatcher/MongoConnect.py:235]: Here you would also need the 
```python
                if phys_det not in phys_stat:
                    phys_stat[phys_det] = []
```
unless you use the defaultdict, right?

File: dispatcher/MongoConnect.py [hunk 3]
```diff
@@ -293,8 +260,28 @@ def aggregate_status(self):
         except Exception as e:
             self.log.error(f'DB snafu? Couldn\'t update aggregate status. '
                             f'{type(e)}, {e}')
+
+        self.physcal_status = phys_stat
```
Comment (jorana) [inline dispatcher/MongoConnect.py:264]: Is this used somewhere, at least I cannot find it?

File: dispatcher/MongoConnect.py [hunk 4]
```diff
@@ -203,6 +198,7 @@ def aggregate_status(self):
                     'pll_unlocks': 0,
                     'number': -1}
                 for k in self.dc}
+        phys_stat = {}
         for detector in self.latest_status.keys():
             # detector = logical
             statuses = {}
```
Comment (jorana) [inline dispatcher/MongoConnect.py:204]: For me personally, it's getting a bit hard to follow all the `status`, `statuses` and the `doc['status']` fields are supposed to do and how they differ

File: dispatcher/dispatcher.py [hunk 5]
```diff
@@ -35,7 +31,8 @@ def main():
     sh = daqnt.SignalHandler()
     SlackBot = daqnt.DaqntBot(os.environ['SLACK_KEY'])
     Hypervisor = daqnt.Hypervisor(control_mc[config['ControlDatabaseName']], logger,
-            daq_config['tpc'], vme_config, sh=sh, testing=args.test, slackbot=SlackBot)
+            daq_config, vme_config, control_inputs=config['ControlKeys'].split(), sh=sh,
+            testing=args.test, slackbot=SlackBot)
     MongoConnector = MongoConnect(config, daq_config, logger, control_mc, runs_mc, Hypervisor, args.test)
     DAQControl = DAQController(config, daq_config, MongoConnector, logger, Hypervisor)
```
Comment (jorana) [inline dispatcher/dispatcher.py:37]: You probably also wanted to add 
`Hypervisor.mongo_connect = MongoConnector`
somewhere here?

File: dispatcher/MongoConnect.py (hunk 4 above)
Comment (jorana) [reply dispatcher/MongoConnect.py:204]: It's probably clearer than it was before by the way so thanks for cleaning it up 👍 

I'm just mentioning this because a mistake here would not be covered by this current review.

**Review by jorana — COMMENTED** [review]
Thanks Darryl!

File: dispatcher/MongoConnect.py (hunk 3 above)
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:264]: This is only used by the hypervisor. It's also misspelled.

File: dispatcher/MongoConnect.py (hunk 4 above)
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:204]: `status` is the extracted status of the host currently being evaluated, or the aggregated status of a logical detector. It's largely a short-term variable. `statuses` is the dict of all the extracted statuses so far, and `doc['status']` is whatever's in the document, which is probably unprocessed.

File: dispatcher/dispatcher.py (hunk 5 above)
Comment (darrylmasson) [reply dispatcher/dispatcher.py:37]: You mean like in the first line below this?

File: dispatcher/MongoConnect.py (hunk 1 above)
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:201]: Or `phys_stat = {k:[] for k in self.dc}`

File: dispatcher/MongoConnect.py (hunk 2 above)
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:235]: This is L210, I guess we could also add it after L226 but there are cleaner solutions.

File: dispatcher/dispatcher.py (hunk 5 above)
Comment (jorana) [reply dispatcher/dispatcher.py:37]: 🤦 , sometimes I oversee things in these tracked changes of git. Thanks

File: dispatcher/MongoConnect.py [hunk 6]
```diff
@@ -289,19 +278,23 @@ def is_timeout(self, doc, t):
         has_ackd = self.host_ackd_command(host)
         ret = False
         if dt > self.timeout:
-            self.log.debug(f'{host} last reported {int(dt)} sec ago')
+            self.logger.debug(f'{host} last reported {int(dt)} sec ago')
             ret = ret or True
         if has_ackd is not None and t - has_ackd > self.timeout_take_action:
-            self.log.critical(f'{host} hasn\'t ackd a command from {int(t-has_ackd)} sec ago')
+            host not in self.host_is_timeout:
```
Comment (jorana) [inline dispatcher/MongoConnect.py:284]: forgot an `if`?

**Review by jorana — COMMENTED** [review]
I just looked at the last commit. I hope I got it all in between the many `log` -> `logger` changes (by the way, the init arg of DAQController is still log)

File: dispatcher/MongoConnect.py (hunk 6 above)
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:284]: Whoops

File: dispatcher/MongoConnect.py [hunk 7]
```diff
@@ -289,19 +278,23 @@ def is_timeout(self, doc, t):
...
                 dt = (now() - self.hv_timeout_fix[host]).total_seconds()
                 if dt > self.hypervisor_host_restart_timeout:
                     self.log_error(f'Hypervisor fixes timeout of {host}', "ERROR", "ERROR")
                     self.hypervisor.handle_timeout(host)
                     self.hv_timeout_fix[host] = now()
                 else:
-                    self.log.debug(f'Not restarting {host}, timeout at {int(dt)}')
+                    self.logger.debug(f'Not restarting {host}, timeout at {int(dt)}')
             ret = ret or True
+        if not ret and host in self.host_is_timeout:
```
[the remark this answers is not in the document]
Comment (darrylmasson) [reply dispatcher/MongoConnect.py:296]: `ret or True` is a boolean aggregation technique that originated in a much earlier version of this function that I never replaced.
